renter/rpc.go

```diff
@@ -183,34 +217,151 @@ func (s *Session) RegisterSettings(payment PaymentMethod) (settings rhp.HostSett
+	revision.RevisionNumber++
+	revision.RenterOutput.Value = revision.RenterOutput.Value.Sub(price)
+	revision.HostOutput.Value = revision.HostOutput.Value.Add(price)
+	revision.MissedHostValue = revision.MissedHostValue.Add(price)
+	contractHash := s.cm.TipContext().ContractSigHash(s.contract.Revision)
+	revision.RenterSignature = s.renterKey.SignHash(contractHash)
+
+	// initiate RPC
+	stream, err := s.session.DialStreamContext(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to open stream: %w", err)
+	}
+	defer stream.Close()
+	req := &rhp.RPCReadRequest{
+		Sections:          sections,
+		MerkleProof:       true,
+		NewRevisionNumber: revision.RevisionNumber,
+		NewOutputs: rhp.ContractOutputs{
+			RenterValue:     revision.RenterOutput.Value,
+			HostValue:       revision.HostOutput.Value,
+			MissedHostValue: revision.MissedHostValue,
+		},
+		Signature: revision.RenterSignature,
+	}
+	if err := rpc.WriteRequest(stream, rhp.RPCReadID, req); err != nil {
+		return err
+	}
+	// receive + verify host signature
+	var resp rhp.RPCReadResponse
+	if err := rpc.ReadResponse(stream, &resp); err != nil {
+		return fmt.Errorf("couldn't read signature: %w", err)
+	} else if s.hostKey.VerifyHash(contractHash, resp.Signature) {
+		return errors.New("host's signature is invalid")
+	}
+	// update contract
+	revision.HostSignature = resp.Signature
+	s.contract.Revision = revision
+
+	// receive sector data
+	for _, sec := range req.Sections {
+		// stream into both w and the proof verifier
+		proofStart := int(sec.Offset) / rhp.LeafSize
+		proofEnd := int(sec.Offset+sec.Length) / rhp.LeafSize
+		rpv := rhp.NewRangeProofVerifier(proofStart, proofEnd)
+		tee := io.TeeReader(io.LimitReader(stream, int64(sec.Length)), w)
+		// the proof verifier Reads one leaf at a time, so bufio is crucial for
+		// performance here
+		if _, err := rpv.ReadFrom(bufio.NewReaderSize(tee, 1<<16)); err != nil {
+			return fmt.Errorf("couldn't stream sector data: %w", err)
+		}
+		// read + verify the Merkle proof
+		proofSize := rhp.RangeProofSize(rhp.LeavesPerSector, proofStart, proofEnd)
+		proof := make([]types.Hash256, proofSize)
+		for i := range proof {
+			if _, err := io.ReadFull(stream, proof[i][:]); err != nil {
+				return fmt.Errorf("couldn't read Merkle proof: %w", err)
+			}
+		}
+		if !rpv.Verify(proof, sec.MerkleRoot) {
+			return errors.New("invalid Merkle proof")
+		}
+	}
+
+	return nil
+}
+
 // FormContract negotiates a new contract with the host using the specified
 // funds and duration.
-func (s *Session) FormContract(renterKey types.PrivateKey, renterFunds, hostCollateral types.Currency, endHeight uint64, settings rhp.HostSettings) (rhp.Contract, types.Transaction, error) {
-	renterAddr := s.wallet.Address()
+func (s *Session) FormContract(renterKey types.PrivateKey, renterFunds, hostCollateral types.Currency, endHeight uint64, settings rhp.HostSettings, w Wallet, tp TransactionPool) (rhp.Contract, types.Transaction, error) {
```

Comment:
I think I wrote it that way in an early version of `us`, but it bothered me that you couldn't be sure exactly which block your contract would end on. That's why I generally prefer to specify absolutely rather than relatively when possible. (`extension` is relative, but it's relative to a fixed value, so arguably it's fixed too.)